packages/editor/src/components/block-manager/index.js

```diff
@@ -17,41 +17,70 @@ import { unlock } from '../../lock-unlock';
 import { store as editorStore } from '../../store';
 import BlockManagerCategory from './category';
 
-function BlockManager( {
-	blockTypes,
-	categories,
-	hasBlockSupport,
-	isMatchingSearchTerm,
-	numberOfHiddenBlocks,
-	enableAllBlockTypes,
-} ) {
+export default function BlockManager() {
 	const debouncedSpeak = useDebounce( speak, 500 );
 	const [ search, setSearch ] = useState( '' );
+	const { showBlockTypes } = unlock( useDispatch( editorStore ) );
 
-	// Filtering occurs here (as opposed to `withSelect`) to avoid
-	// wasted renders by consequence of `Array#filter` producing
-	// a new value reference on each call.
-	blockTypes = blockTypes.filter(
-		( blockType ) =>
-			hasBlockSupport( blockType, 'inserter', true ) &&
-			( ! search || isMatchingSearchTerm( blockType, search ) ) &&
-			( ! blockType.parent ||
-				blockType.parent.includes( 'core/post-content' ) )
-	);
+	const {
+		blockTypes,
+		categories,
+		hasBlockSupport,
+		isMatchingSearchTerm,
+		numberOfHiddenBlocks,
+	} = useSelect( ( select ) => {
+		// Some hidden blocks become unregistered
+		// by removing for instance the plugin that registered them, yet
+		// they're still remain as hidden by the user's action.
+		// We consider "hidden", blocks which were hidden and
+		// are still registered.
+		const _blockTypes = select( blocksStore ).getBlockTypes();
+		const hiddenBlockTypes = (
+			select( preferencesStore ).get( 'core', 'hiddenBlockTypes' ) ?? []
+		).filter( ( hiddenBlock ) => {
+			return _blockTypes.some(
+				( registeredBlock ) => registeredBlock.name === hiddenBlock
+			);
+		} );
+
+		return {
+			blockTypes: _blockTypes,
+			categories: select( blocksStore ).getCategories(),
+			hasBlockSupport: select( blocksStore ).hasBlockSupport,
+			isMatchingSearchTerm: select( blocksStore ).isMatchingSearchTerm,
+			numberOfHiddenBlocks:
+				Array.isArray( hiddenBlockTypes ) && hiddenBlockTypes.length,
+		};
+	}, [] );
+
+	function enableAllBlockTypes( newBlockTypes ) {
+		const blockNames = newBlockTypes.map( ( { name } ) => name );
+		showBlockTypes( blockNames );
+	}
+
+	const filteredBlockTypes = useMemo( () => {
```

Comment:
I have taken [this comment](https://github.com/WordPress/gutenberg/pull/65349/files#diff-46eb4f22bc67296bbbdc66ab3fc638171702e900fb54cae22a37d636bb28cd09L31-L33) into account and cached the results, but is this a good idea?

Reply:
The original comment concerns returning a new array reference from the `mapSelect` callback. This would result in a rerender every time the block store is updated and a warning from `useSelect` in dev mode.

It should be okay to leave it without memoization. But don't have a strong opinion here.

Reply:
I see. Just to be safe, I would like to remove `useMemo` and then merge.

Reply:
Sounds good 👍